builtin/providers/aws/resource_aws_lambda_function.go

```diff
@@ -297,19 +297,34 @@ func resourceAwsLambdaFunctionUpdate(d *schema.ResourceData, meta interface{}) e
 	}
 
 	codeUpdate := false
-	if v, ok := d.GetOk("filename"); ok && d.HasChange("source_code_hash") {
-		file, err := loadFileContent(v.(string))
-		if err != nil {
-			return fmt.Errorf("Unable to load %q: %s", v.(string), err)
+	if v, ok := d.GetOk("filename"); ok {
+		if d.HasChange("source_code_hash") {
+			file, err := loadFileContent(v.(string))
+			if err != nil {
+				return fmt.Errorf("Unable to load %q: %s", v.(string), err)
+			}
+			codeReq.ZipFile = file
+			codeUpdate = true
+		}
+	} else {
+		s3Bucket, bucketOk := d.GetOk("s3_bucket")
+		s3Key, keyOk := d.GetOk("s3_key")
+		s3ObjectVersion, versionOk := d.GetOk("s3_object_version")
+
+		if bucketOk && keyOk {
+			if d.HasChange("s3_bucket") || d.HasChange("s3_key") {
+				codeReq.S3Bucket = aws.String(s3Bucket.(string))
+				codeReq.S3Key = aws.String(s3Key.(string))
```

Comment:
Yes, you're right. IMO `s3_bucket` and `s3_key` must be set for every request, even though they haven't changed.

Your pseudocode is more readable than my code, I will update the PR to reflect it.